node/toolrunner.ts

```diff
@@ -580,200 +580,234 @@ export class ToolRunner extends events.EventEmitter {
-                defer.reject(error);
-            }
-        });
+            let cp = child.spawn(this._getSpawnFileName(), this._getSpawnArgs(options), this._getSpawnOptions(options));
 
-        cp.on('close', (code, signal) => {
-            waitingEvents--; //process is complete
-            this._debug('rc:' + code);
-            returnCode = code;
+            var stdbuffer: string = '';
+            cp.stdout.on('data', (data: Buffer) => {
+                this.emit('stdout', data);
 
-            if (stdbuffer.length > 0) {
-                this.emit('stdline', stdbuffer);
-            }
+                if (!options.silent) {
+                    options.outStream.write(data);
+                }
 
-            if (errbuffer.length > 0) {
-                this.emit('errline', errbuffer);
-            }
+                this._processLineBuffer(data, stdbuffer, (line: string) => {
+                    this.emit('stdline', line);
+                });
+            });
 
-            if (code != 0 && !options.ignoreReturnCode) {
-                success = false;
-            }
+            var errbuffer: string = '';
+            cp.stderr.on('data', (data: Buffer) => {
+                state.processStderr = true;
+                this.emit('stderr', data);
 
-            this._debug('success:' + success);
-                
-            if (!successFirst) { //in the case output is piped to another tool, check exit code of both tools
-                error = new Error(toolPathFirst + ' failed with return code: ' + returnCodeFirst);
-            } else if (!success) {
-                error = new Error(toolPath + ' failed with return code: ' + code);
-            }
+                if (!options.silent) {
+                    var s = options.failOnStdErr ? options.errStream : options.outStream;
+                    s.write(data);
+                }
+
+                this._processLineBuffer(data, errbuffer, (line: string) => {
+                    this.emit('errline', line);
+                });
+            });
 
-            if(waitingEvents == 0) {
-                if (error) {
-                    defer.reject(error);
-                } else {
-                    defer.resolve(returnCode);
+            cp.on('error', (err) => {
+                state.processError = err.message;
+                state.processExited = true;
+                state.processClosed = true;
+                state.CheckComplete();
+            });
+
+            cp.on('exit', (code, signal) => {
+                state.processExitCode = code;
+                state.processExited = true;
+                this._debug(`Exit code ${code} received from tool '${this.toolPath}'`);
+                state.CheckComplete()
+            });
+
+            cp.on('close', (code, signal) => {
+                state.processExitCode = code;
+                state.processClosed = true;
+                this._debug(`STDIO streams have closed for tool '${this.toolPath}'`)
```

Comment:
do we also want to set state.processExited = true; in case we miss the exit event.